classes/file/PKPLibraryFileManager.php

```diff
@@ -112,16 +112,49 @@ public function &copyFromTemporaryFile(&$temporaryFile, $libraryFileType)
         $libraryFileDao = DAORegistry::getDAO('LibraryFileDAO'); /** @var LibraryFileDAO $libraryFileDao */
         $libraryFile = $libraryFileDao->newDataObject();
 
+        $libraryFile = $this->assignFromTemporaryFile($temporaryFile, $libraryFileType, $libraryFile);
+        if (!$this->copyFile($temporaryFile->getFilePath(), $this->getBasePath() . $libraryFile->getServerFileName())) {
+            return false;
+        }
+
+        return $libraryFile;
+    }
+
+    /**
+     * Routine to replace a library file from a temporary file.
+     * @param $temporaryFile object
+     * @param $libraryFileType int LIBRARY_FILE_TYPE_...
+     * @param $libraryFile LibraryFile 
+     * @return LibraryFile|false the updated LibraryFile, or false on error
+     */
+    function &replaceFromTemporaryFile(&$temporaryFile, $libraryFileType, $libraryFile) {
+        $originalServerFilename = $libraryFile->getServerFileName();
+
+        $libraryFile = $this->assignFromTemporaryFile($temporaryFile, $libraryFileType, $libraryFile);
+        if (!$this->copyFile($temporaryFile->getFilePath(), $this->getBasePath() . $libraryFile->getServerFileName())) {
+                return false;
+        }
+
+        if ($originalServerFilename !== $libraryFile->getServerFileName()) {
+                unlink($this->getBasePath() . $originalServerFilename);
+        }
+        return $libraryFile;
+    }
+
+    /**
+     * Routine to assign metadata to a library file from a temporary file
+     * @param $temporaryFile object
+     * @param $libraryFileType int LIBRARY_FILE_TYPE_...
+     * @param $libraryFile LibraryFile 
+     * @return LibraryFile the updated LibraryFile
+     */
+    function &assignFromTemporaryFile(&$temporaryFile, $libraryFileType, $libraryFile) {
         $libraryFile->setDateUploaded($temporaryFile->getDateUploaded());
         $libraryFile->setDateModified($temporaryFile->getDateUploaded());
         $libraryFile->setFileType($temporaryFile->getFileType());
         $libraryFile->setFileSize($temporaryFile->getFileSize());
         $libraryFile->setServerFileName($this->generateFileName($libraryFileType, $temporaryFile->getOriginalFileName()));
         $libraryFile->setOriginalFileName($temporaryFile->getOriginalFileName());
-        if (!$this->copyFile($temporaryFile->getFilePath(), $this->getBasePath() . $libraryFile->getServerFileName())) {
```

Comment:
Could you write a little bit about this change? At a glance, I think it might have unintended consequences.

Reply:
Previously the `copyFromTemporaryFile()` method handled both the metadata assignment and the file copy operation.  Breaking this into `copyFromTemporaryFile()` and `replaceFromTemporaryFile()`, we move the metadata operations into their own method.  It didn't seem appropriate for the `assignFromTemporaryFile()` to perform the copy, so each of `copyFromTemporaryFile()` (line 116) and `replaceFromTemporaryFile()` (line 134) to this themselves.

Reply:
That makes sense, thanks!